src/vm/ast/parser/mod.rs

```diff
@@ -282,22 +295,57 @@ pub fn parse_lexed(mut input: Vec<(LexItem, u32, u32)>) -> ParseResult<Vec<PreSy
-                        },
-                        Some((ref mut list, _, _)) => {
-                            list.push(pre_expr);
-                        }
-                    };
+                if let Some((value, start_line, start_column, expected_token)) = parse_stack.pop() {
+                    if let LexItem::RightParen = expected_token {
+                        let mut pre_expr = PreSymbolicExpression::list(value.into_boxed_slice());
+                        pre_expr.set_span(start_line, start_column, line_pos, column_pos);
+                        match parse_stack.last_mut() {
+                            None => {
+                                // no open lists on stack, add current to result.
+                                output_list.push(pre_expr)
+                            },
+                            Some((ref mut list, _, _, _)) => {
+                                list.push(pre_expr);
+                            }
+                        };
+                    } else {
+                        return Err(ParseError::new(ParseErrors::ClosingParenthesisExpected))
+                    }
+                } else {
+                    return Err(ParseError::new(ParseErrors::ClosingParenthesisUnexpected))
+                }
+            },
+            LexItem::LeftCurly => {
+                let new_list = Vec::new();
+                parse_stack.push((new_list, line_pos, column_pos, LexItem::RightCurly));
+            },
+            LexItem::RightCurly => {
+                if let Some((value, start_line, start_column, expected_token)) = parse_stack.pop() {
+                    if let LexItem::RightCurly = expected_token {
+                        let mut pairs = value.chunks(2)
+                                         .map(|pair| pair.to_vec().into_boxed_slice())
+                                         .map(PreSymbolicExpression::list)
+                                         .collect::<Vec<_>>();
+                        pairs.insert(0, PreSymbolicExpression::atom("tuple".to_string().try_into().unwrap()));
+                        let mut pre_expr = PreSymbolicExpression::list(pairs.into_boxed_slice());
+                        pre_expr.set_span(start_line, start_column, line_pos, column_pos);
+                        match parse_stack.last_mut() {
+                            None => {
+                                // no open lists on stack, add current to result.
+                                output_list.push(pre_expr)
+                            },
+                            Some((ref mut list, _, _, _)) => {
+                                list.push(pre_expr);
+                            }
+                        };
+                    } else {
+                        // FIX: Use ParseErrors::ClosingTupleLiteralExpected
+                        return Err(ParseError::new(ParseErrors::ClosingParenthesisExpected))
+                    }
```

Comment:
This is moving in the right direction :). For expanding this curly brace sugared syntax, instead of generating some code, I think I would:  
1) Create a new type of PreSymbolicExpression
https://github.com/blockstack/blockstack-core/blob/7d162419f0bd77a4273a158a989669ee5f8fe9c0/src/vm/representations.rs#L75-L83
2) When parsing, be generating a new instance of the aforementioned PreSymbolicExpressionType
3) In the SugarExpander pass
 https://github.com/blockstack/blockstack-core/blob/7d162419f0bd77a4273a158a989669ee5f8fe9c0/src/vm/ast/sugar_expander/mod.rs#L41
Matching the pre-symbolic-expression and generating a `SymbolicExpression::literal_value`, holding the tuple.

Reply:
@lgalabru Thanks for the pointer. Implemented sugar expander in 52467a4 using a new `PreSymbolicExpression::Tuple`. Can `SymbolicExpression::literal_value` be used to hold a tuple where the value expression is evaluated?

Reply:
Getting close?

Reply:
Good point @njordhov -- you can't use `literal_value` for runtime-evaluated tuple values. I think the current approach of expanding to an explicit `(tuple ...)` call is the right way to do it.